nntrainer/src/databuffer_v2.cpp

```diff
@@ -0,0 +1,325 @@
+    }
+
+    /**
+     * Fill the data into inputs and labels, and remove from batch buffers
+     * decrement avail buffer idx
+     */
+    if (data_available) {
+      // TODO: remove the front element first, so that background threads have
+      // to wait less
+      /**
+       * set inputs and labels data from batched buffer_elem
+       * As this batch is already completed, background thread will not change
+       * this data. So no batch lock is needed here. No thread lock is needed
+       * here as the data buffer caller is our thread and it is assumed that
+       * when getData() is being called, stop() wont be called.
+       */
+      auto batched_buffer_elem = batched_buffer.front();
+      void **batched_inputs = std::get<0>(batched_buffer_elem);
+      for (unsigned int idx = 0; idx < inputs.size(); idx++)
+        inputs[idx]->setData(static_cast<float *>(batched_inputs[idx]));
+
+      void **batched_labels = std::get<1>(batched_buffer_elem);
+      for (unsigned int idx = 0; idx < labels.size(); idx++)
+        labels[idx]->setData(static_cast<float *>(batched_labels[idx]));
+
+      /** Remove the front consumed data from the buffer */
+      {
+        std::lock(buffer_m, batched_buffer_m);
+        std::lock_guard<std::mutex> buffer_data_avail_lock(buffer_m,
+                                                           std::adopt_lock);
+        std::lock_guard<std::mutex> batched_buffer_data_avail_lock(
+          batched_buffer_m, std::adopt_lock);
+        batched_buffer.pop_front();
+        for (unsigned int idx = 0; idx < batch_size; idx++)
+          buffer.pop_front();
+        avail_buffer_idx -= batch_size;
+
+        if (avail_buffer_idx < 0 || avail_buffer_idx > buffer_len)
+          throw std::logic_error("Buffer overflow error");
+      }
+
+      /** Add new space to the buffer */
+      pushBatchedData();
+      /** Wake up background threads waiting for more space to fill data */
+      buffer_cond_hungry.notify_all();
+      return true;
+    }
+
+    // TODO: optimize to wake up only when full batch is filled by the
+    // background thread
+    /** wait for background thread to fill the data */
+    std::unique_lock<std::mutex> batched_buffer_wait_for_data(batched_buffer_m);
+    buffer_cond_filled.wait(batched_buffer_wait_for_data);
+  }
+
+  throw std::logic_error("Reach unexpected code location");
+}
+
+void DataBuffer_v2::collectData() {
+  bool last_element = false;
+  while (!last_element) {
+    bool buffer_space_available;
+    std::tuple<void **, void **> buffer_elem;
+
+    /**
+     * THREAD_REQUEST_TO_STOP : requested by main thread to stop
+     * THREAD_ERROR : thread is in error state, exit
+     * THREAD_EPOCH_FINISHED : thread finished work, exit
+     */
+    {
+      std::lock_guard<std::mutex> thread_stop(thread_m);
+      if (thread_state != ThreadStates::THREAD_RUNNING) {
+        thread_state = ThreadStates::THREAD_STOPPED;
```

Comment:
THREAD_ERROR will be overridden here. Is this possible scenario and is that okay?

Reply:
Yes, in case of multiple background threads, it will be overridden. 
Updated.

Reply:
Actually there is no much use for THREAD_REQUEST_TO_STOP and THREAD_STOPPED state.
Merged into THREAD_JOINABLE as per https://github.com/nnstreamer/nntrainer/pull/305/files#r484586941.